src/Lecture2.hs

```diff
@@ -1,170 +1,215 @@
+  | otherwise = x * lazyProduct xs
+
+-- | Implement a function that duplicates every element in the list.
+--
+-- >>> duplicate [3, 1, 2]
+-- [3,3,1,1,2,2]
+-- >>> duplicate "cab"
+-- "ccaabb"
 duplicate :: [a] -> [a]
-duplicate = error "TODO"
-
-{- | Implement function that takes index and a list and removes the
-element at the given position. Additionally, this function should also
-return the removed element.
-
->>> removeAt 0 [1 .. 5]
-(Just 1,[2,3,4,5])
-
->>> removeAt 10 [1 .. 5]
-(Nothing,[1,2,3,4,5])
--}
-removeAt = error "TODO"
-
-{- | Write a function that takes a list of lists and returns only
-lists of even lengths.
-
->>> evenLists [[1,2,3], [3,1,2,7], [], [5, 7, 2]]
-[[3,1,2,7],[]]
-
-♫ NOTE: Use eta-reduction and function composition (the dot (.) operator)
-  in this function.
--}
-evenLists = error "TODO"
+duplicate [] = []
+duplicate (x : xs) = x : x : duplicate xs
+
+-- | Implement function that takes index and a list and removes the
+-- element at the given position. Additionally, this function should also
+-- return the removed element.
+--
+-- >>> removeAt 0 [1 .. 5]
+-- (Just 1,[2,3,4,5])
+--
+-- >>> removeAt 10 [1 .. 5]
+-- (Nothing,[1,2,3,4,5])
+removeAt :: Int -> [a] -> (Maybe a, [a])
+removeAt x list
+  | x < 0 = (Nothing, list)
+  | null list = (Nothing, list)
+  | otherwise = case x of
+      0 -> (Just (head list), tail list)
+      _ ->
+        let flist = take x list
+            blist = drop x list
+         in getList flist blist
+        where
+          getList :: [a] -> [a] -> (Maybe a, [a])
+          getList l1 [] = (Nothing, l1)
+          getList l1 l2 = (Just (head l2), l1 ++ tail l2)
+
+-- | Write a function that takes a list of lists and returns only
+-- lists of even lengths.
+--
+-- >>> evenLists [[1,2,3], [3,1,2,7], [], [5, 7, 2]]
+-- [[3,1,2,7],[]]
+--
+-- ♫ NOTE: Use eta-reduction and function composition (the dot (.) operator)
+--  in this function.
+evenLists :: [[a]] -> [[a]]
+evenLists = filter (even . length)
+
+-- | The @dropSpaces@ function takes a string containing a single word
+-- or number surrounded by spaces and removes all leading and trailing
+-- spaces.
+--
+-- >>> dropSpaces "   hello  "
+-- "hello"
+-- >>> dropSpaces "-200            "
+-- "-200"
+--
+-- ♫ NOTE: As in the previous task, use eta-reduction and function
+--  composition (the dot (.) operator) in this function.
+--
+-- 🕯 HINT: look into Data.Char and Prelude modules for functions you may use.
+
+-- Fails for infinite whitespaces at the end
+-- dropSpaces = filter (not . isSpace)
+
+-- Not the most ideal solution
+-- But takes care of the infinite whitespaces at the end
+dropSpaces :: String -> String
+dropSpaces = dropc []
+  where
+    dropc :: String -> String -> String
+    dropc [] [] = []
+    dropc w1 [] = reverse w1
+    dropc [] (w : w2)
+      | isSpace w = dropc [] w2
+      | otherwise = dropc [w] w2
+    dropc w1 (w : w2)
+      | isSpace w = reverse w1
+      | otherwise = dropc (w : w1) w2
```

Comment:
That's the correct solution 👍🏻 

You can also use the standard function [dropWhile](https://hackage.haskell.org/package/base-4.17.0.0/docs/Prelude.html#v:dropWhile) and [takeWhile](https://hackage.haskell.org/package/base-4.17.0.0/docs/Prelude.html#v:takeWhile) to the implement the solution shorter.